core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -283,32 +284,55 @@ public Snapshot apply() {
       throw new RuntimeIOException(e, "Failed to write manifest list file");
     }
 
+    Map<String, String> summary = summary();
+    String operation = operation();
+
     Long addedRows = null;
-    Long lastRowId = null;
-    if (base.rowLineageEnabled()) {
-      addedRows = calculateAddedRows(manifests);
-      lastRowId = base.nextRowId();
+    Long firstRowId = null;
+    if (base.formatVersion() >= 3) {
+      addedRows = calculateAddedRows(operation, summary, manifests);
+      firstRowId = base.nextRowId();
     }
 
     return new BaseSnapshot(
         sequenceNumber,
         snapshotId(),
         parentSnapshotId,
         System.currentTimeMillis(),
-        operation(),
-        summary(base),
+        operation,
+        summaryWithTotals(base, summary),
         base.currentSchemaId(),
         manifestList.location(),
-        lastRowId,
+        firstRowId,
         addedRows);
   }
 
-  private Long calculateAddedRows(List<ManifestFile> manifests) {
+  private Long calculateAddedRows(
+      String operation, Map<String, String> summary, List<ManifestFile> manifests) {
+    if (summary != null) {
+      long addedRecords =
+          PropertyUtil.propertyAsLong(summary, SnapshotSummary.ADDED_RECORDS_PROP, 0L);
```

Comment:
When the summary is defined, we trust the added-records and deleted-records values, which should match the counts calculated below.

I also added a check to handle the `DataOperations.REPLACE` case. For a replace, the added records should always be 0. If we want, we can also add checks for data file metadata that assert there are no null values for the `_row_id` column. What do you think @RussellSpitzer?

Reply:
I've actually been thinking about this a bit more and I'm not sure we can do this. If Summary != null we can't assume that Added_Records_Prop is defined. I think if it is null we probably should still fall back to manifest inspection

Reply:
We can trust the summary to be correct given that the operation produces it. However, I think it is best not to rely on this and to use actual rows assigned by the writers instead. I'll follow up with a fix.